File: 2018/day25.rs
use std::collections::{HashSet, VecDeque};
use std::fs::read_to_string;

fn main() {
    let text = read_to_string("/home/xdavidliu/Documents/temp/data.txt").unwrap();
    let mut points = Vec::new();
    for line in text.lines() {
        let mut words = line.split(',');
        let mut xs: Vec<i32> = Vec::new();
        for _ in 0..4 {
            xs.push(words.next().unwrap().parse().unwrap());
        }
        points.push((xs[0], xs[1], xs[2], xs[3]));
    }
    let mut adj = vec![vec![]; points.len()];
    for i in 0..points.len()-1 {
        for k in i+1..points.len() {
            if near(&points[i], &points[k]) {
                adj[i].push(k);
                adj[k].push(i);
            }
        }
    }
    let mut seen = HashSet::new();
    let mut components = 0;
    for i in 0..points.len() {
        if seen.contains(&i) { continue; }
        let mut que = VecDeque::new();
        que.push_back(i);
        seen.insert(i);
        while let Some(k) = que.pop_front() {
            for &m in &adj[k] {
                if !seen.contains(&m) {
                    seen.insert(m);
                    que.push_back(m);
                }
            }
        }
        components += 1;
    }
    println!("part 1 = {components}");  // 375
}

fn near(me: &(i32, i32, i32, i32), other: &(i32, i32, i32, i32)) -> bool {
    let mut dist = 0;
    dist += (me.0 - other.0).abs();
    dist += (me.1 - other.1).abs();
    dist += (me.2 - other.2).abs();
    dist += (me.3 - other.3).abs();
    dist <= 3
}

File: 2023-cpp/day25.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>
#include <set>
#include <deque>
#include <map>

// alternate proper way to do this:
// https://stackoverflow.com/questions/4482986/how-can-i-find-the-minimum-cut-on-a-graph-using-a-maximum-flow-algorithm
// also, see ch 26 min cut max flow section in CLRS

void adj_insert(std::map<std::string, std::vector<std::string>> &adj,
                const std::string &key, const std::string &val) {
    auto [iter, ignore] = adj.insert({key, std::vector<std::string>()});
    // "Each connection between two components is represented only once"
    iter->second.push_back(val);
}

auto read_adj() {
    const char *filepath = "/home/employee/Documents/temp/data.txt";
    auto fs = std::ifstream(filepath);
    std::map<std::string, std::vector<std::string>> adj;
    std::string line, key, val;
    while (std::getline(fs, line)) {
        std::istringstream iss(line);
        iss >> key;
        key = key.substr(0, key.size() - 1);  // omit :
        while (iss >> val) {
            adj_insert(adj, key, val);
            adj_insert(adj, val, key);
        }
    }
    return adj;
}

std::pair<std::string, std::string> ordered(const std::string &one, const std::string &two) {
    if (one < two) { return {one, two}; }
    else { return {two, one}; }
}

void increment(std::map<std::pair<std::string, std::string>, int> &count,
               const std::string &one,
               const std::string &two)
{
    const auto key = ordered(one, two);
    auto [iter, ignore] = count.insert({key, 0});
    ++iter->second;
}

void bfs(const std::map<std::string, std::vector<std::string>> &adj,
         const std::string &start,
         std::set<std::string> &seen,
         std::map<std::pair<std::string, std::string>, int> &count)
{
    std::deque<std::string> que;
    que.push_back(start);
    seen.insert(start);
    while (!que.empty()) {
        const auto front = que.front();
        que.pop_front();
        for (const auto &other : adj.at(front)) {
            if (!seen.count(other)) {
                seen.insert(other);
                increment(count, front, other);
                que.push_back(other);
            }
        }
    }
}

void remove_one_edge(std::map<std::string, std::vector<std::string>> &adj,
                 const std::string &key,
                 const std::string &elem)
{
    auto &v = adj.at(key);
    v.erase(std::find(v.begin(), v.end(), elem));
}

void remove_both_edge(std::map<std::string, std::vector<std::string>> &adj,
                      const std::string &one,
                      const std::string &two)
{
    remove_one_edge(adj, one, two);
    remove_one_edge(adj, two, one);
}

auto top_key(const std::map<std::pair<std::string, std::string>, int> &count) {
    int best = 0;
    std::string one, two;
    for (const auto &[key, c] : count) {
        if (c > best) {
            best = c;
            one = key.first;
            two = key.second;
        }
    }
    return std::make_pair(one, two);
}

void foo1() {
    auto adj = read_adj();
    std::map<std::pair<std::string, std::string>, int> count;
    std::set<std::string> seen;
    for (int i = 0; i < 3; ++i) {
        for (const auto &[key, ignore] : adj) {
            bfs(adj, key, seen, count);
            seen.clear();
        }
        const auto [one, two] = top_key(count);
        remove_both_edge(adj, one, two);
        count.clear();
    }
    int seen_size = 0;
    std::vector<int> diffs;
    for (const auto &[key, ignore] : adj) {
        bfs(adj, key, seen, count);
        if (seen.size() > seen_size) {
            diffs.push_back(seen.size() - seen_size);
            seen_size = seen.size();
        }
    }
    std::cout << "part 1 = " << (diffs[0] * diffs[1]) << '\n';
    // 547410
    /*
     * start BFS from each visited edge
     * update counter, after done, sort by count, get top 5
     * if yes, then cut those and do bfs from each component and multiply
     *
     * update: do it three times, taking top each time
     *
     */
}

int main() {
    foo1();
}
